run-tests.php

```diff
@@ -1349,71 +1348,29 @@ function run_all_tests_parallel($test_files, $env, $redir_tested) {
 	echo "====⚡️==== WELCOME TO THE FUTURE: run-tests PARALLEL EDITION ====⚡️====\n";
 	echo "====⚡️===========================================================⚡️====\n";
 
-	// Because some of the PHP test suite has not been written with
-	// parallel execution in mind, it is not safe to just run any two tests
-	// concurrently.
-	// Therefore, we divide the test set into directories and test multiple
-	// directories at once, but not multiple tests within them.
-
-	$testDirsToGo = [];
-
+	// Each test may specify a list of conflict keys. While a test that conflicts with
+	// key K is running, no other test that conflicts with K may run. Conflict keys are
+	// specified either in the --CONFLICTS-- section, or @CONFLICTS file inside a directory.
+	$dirConflictsWith = [];
+	$fileConflictsWith = [];
 	foreach ($test_files as $file) {
-		$dirSeparator = strrpos($file, DIRECTORY_SEPARATOR);
-		if ($dirSeparator !== FALSE) {
-			$testDirsToGo[substr($file, 0, $dirSeparator)][] = $file;
-		} else {
-			$testDirsToGo[""][] = $file;
+		$contents = file_get_contents($file);
+		$conflicts = [];
+		if (preg_match('/--CONFLICTS--(.+?)--/', $contents, $matches)) {
+			$conflicts = array_map('trim', explode("\n", trim($matches[1])));
 		}
-	}
 
-	// We assume most test directories should be executed in serial, but for
-	// big directories, this would waste time if they can actually be parallel.
-	// Therefore, if a directory has a special '@CAN_BE_PARALLELISED' file, we
-	// will divide it up into smaller “directories” automatically.
-	foreach ($testDirsToGo as $dir => $tests) {
-		if (count($tests) < 64 || !is_string($dir)) {
-			continue;
-		}
-		if (file_exists($dir . DIRECTORY_SEPARATOR . '@CAN_BE_PARALLELISED')) {
-			foreach (array_chunk($tests, 64) as $testsChunk) {
-				$testDirsToGo[] = $testsChunk;
+		$dir = dirname($file);
+		if (!isset($dirConflictsWith[$dir])) {
+			$dirConflicts = [];
+			if (file_exists($dir . '/@CONFLICTS')) {
```

Comment:
Maybe save this stat-call and better directly try a `file_get_contents` while handling the false-return case?

Reply:
This is done only once per test directory. The bottleneck here is going to be the per-file check above (on Windows).